### ISSUE TITLE
Time2Vec: Learning a Vector Representation of Time

### ISSUE THREAD

**Issue by ntakouris** [tensorflow/addons#2508]
**Relevant information**
- Are you willing to contribute it (yes/no): **yes**
- Are you willing to maintain it going forward? (yes/no): **yes** 
- Is there a relevant academic paper? (if so, where): [Time2Vec: Learning a Vector Representation of Time](https://arxiv.org/abs/1907.05321)
- Does the relavent academic paper exceed 50 citations? (yes/no): **no** (google scholar counts 15)
- Is there already an implementation in another framework? (if so, where): There are several[ like this in pytorch ](https://github.com/ojus1/Time2Vec-PyTorch) and [this in tensorflow ](https://github.com/email81227/Time2Vec-TensorFlow2) but they are lacking: learning supporting multiple features and/or applying the layer in tensors of shape (batch size, seq len, feat size) is not supported.
- Was it part of tf.contrib? (if so, where): **no**

**Which API type would this fall under (layer, metric, optimizer, etc.)**
Layer

**Who will benefit with this feature?**
Primarily time series and other sequential tasks

**Any other info.**
Sample (working properly, but feature-incomplete) implementation:

```py
import tensorflow.keras as keras
import tensorflow.keras.backend as K


class Time2Vec(keras.layers.Layer):
    def __init__(self, kernel_size):
        super().__init__(trainable=True, name='Time2Vec')
        self.k = kernel_size

    def build(self, input_shape):
        # trend
        self.wb = self.add_weight(name='wb', shape=(
            input_shape[1],), initializer='uniform', trainable=True)
        self.bb = self.add_weight(name='bb', shape=(
            input_shape[1],), initializer='uniform', trainable=True)
        # periodic
        self.wa = self.add_weight(name='wa', shape=(
            1, input_shape[1], self.k), initializer='uniform', trainable=True)

        self.ba = self.add_weight(name='ba', shape=(
            1, input_shape[1], self.k), initializer='uniform', trainable=True)
        super().build(input_shape)

    def call(self, inputs, **kwargs):
        bias = self.wb * inputs + self.bb
        dp = K.dot(inputs, self.wa) + self.ba
        wgts = K.sin(dp)

        ret = K.concatenate([K.expand_dims(bias, -1), wgts], -1)
        ret = K.reshape(ret, (-1, inputs.shape[1]*(self.k+1)))
        return ret

    def compute_output_shape(self, input_shape):
        return (input_shape[0], input_shape[1]*(self.k + 1))
```

In order to use that as-is, you'd have to do:

```py
x = tf.random.normal([8, 7, 5]) # bs, seq len, feat

time_embedding = keras.layers.TimeDistributed(Time2Vec(kernel_size=3))(x)

y = K.concatenate([x, time_embedding], -1) # optionally concatenate with input
```

There are some design considerations:

- Are we going to use 2 kernel sizes?, one for the trend and one for the periodic component, or just keep one kernel size and optionally also provide a second one that takes the value of the first one if it is `None`
- Is the `TimeDistributed` part going to be automatically included in the layer by inspecting the `input_shape`?

**Comment by AakashKumarNain** [comment]
Thanks for the relevant information and the sample but there aren't enough citations. We will wait for some more time to check if the number of citation grows over time. If it does, you can reopen the issue and submit a PR.

**Comment by s-g-mo** [comment]
It's been a few years since this original post. The paper indeed has over 50 citations now. Can we re-open this? I'm also trying to adapt this layer for multiple features.
